sw/airborne/modules/computer_vision/opticflow_module.c

```diff
@@ -0,0 +1,225 @@
+ * Boston, MA 02111-1307, USA.
+ */
+
+/**
+ * @file paparazzi/sw/ext/ardrone2_vision/modules/OpticFlow/opticflow_module.h
+ * @brief optical-flow based hovering for Parrot AR.Drone 2.0
+ *
+ * Sensors from vertical camera and IMU of Parrot AR.Drone 2.0
+ */
+
+
+// Own header
+#include "opticflow_module.h"
+
+// Navigate Based On Vision
+#include "opticflow/hover_stabilization.h"
+
+// Paparazzi
+#include "state.h" // for attitude
+#include "boards/ardrone/navdata.h" // for ultrasound Height
+
+// Threaded computer vision
+#include <pthread.h>
+
+// Frame Rate (FPS)
+#include <sys/time.h>
+float FPS;
+volatile long timestamp;
+#define USEC_PER_SEC 1000000L
+long time_elapsed(struct timeval *t1, struct timeval *t2)
+{
+  long sec, usec;
+  sec = t2->tv_sec - t1->tv_sec;
+  usec = t2->tv_usec - t1->tv_usec;
+  if (usec < 0) {
+    --sec;
+    usec = usec + USEC_PER_SEC;
+  }
+  return sec * USEC_PER_SEC + usec;
+}
+
+struct timeval start_time;
+struct timeval end_time;
+
+void start_timer()
+{
+  gettimeofday(&start_time, NULL);
+}
+long end_timer()
+{
+  gettimeofday(&end_time, NULL);
+  return time_elapsed(&start_time, &end_time);
+}
+
+void opticflow_module_init(void)
+{
+  // Immediately start the vision thread when the module initialized
+  opticflow_module_start();
+
+  // Stabilization Code Initialization
+  init_hover_stabilization_onvision();
+
+  // Frame Rate Initialization
+  FPS = 0.0;
+  timestamp = 0;
+  start_timer();
+}
+
+
+volatile uint8_t computervision_thread_has_results = 0;
+
+void opticflow_module_run(void)
+{
+
+  // Read Latest Vision Module Results
+  if (computervision_thread_has_results) {
+    computervision_thread_has_results = 0;
+    run_hover_stabilization_onvision();
+  }
+}
+
+#include "firmwares/rotorcraft/navigation.h"
+
+void guidance_module_run(bool_t inflight)
+{
+  nav_flight_altitude = -1;
+}
+
+
+/////////////////////////////////////////////////////////////////////////
+// COMPUTER VISION THREAD
+
+// Video
+#include "v4l/video.h"
+#include "resize.h"
+
+// Payload Code
+#include "opticflow/opticflow_code.h"
+
+// Downlink Video
+//#define DOWNLINK_VIDEO 1
+
+#ifdef DOWNLINK_VIDEO
+#include "encoding/jpeg.h"
+#include "encoding/rtp.h"
+#endif
+
+#include <stdio.h>
+
+pthread_t computervision_thread;
+volatile uint8_t computervision_thread_status = 0;
+volatile uint8_t computer_vision_thread_command = 0;
+void *computervision_thread_main(void *data);
+void *computervision_thread_main(void *data)
+{
+  // Video Input
+  struct vid_struct vid;
+  vid.device = (char *)"/dev/video2"; // video1 = front camera; video2 = bottom camera
+  vid.w = 320;
+  vid.h = 240;
+  vid.n_buffers = 4;
+
+  if (video_init(&vid) < 0) {
+    printf("Error initialising video\n");
+    computervision_thread_status = -1;
+    return 0;
+  }
+
+  // Video Grabbing
+  struct img_struct *img_new = video_create_image(&vid);
+
+  // Video Resizing
+#define DOWNSIZE_FACTOR 1
+  struct img_struct small;
```

Comment:
if you don't plan to use DOWNSIZE_FACTOR higher than 1, you should probably remove completely `small` variable